MH-12730, Workflow Index Rebuild Performance

lkiesow opened the pull request:
For a workflow-index rebuild Opencast loads and deserializes far more
data that it actually needs. This may even lead to a system needing
hours before it can actually start the index rebuild or for the process
to fail due to the massive amount of required memory.

This patch tries to improve this behavior by requesting the specific
resources directly from the database and deserializing only what is
necessary.

Additionally, it will also avoid sending those messages via message
broker which would only be used for creating log messages. Instead, the
logging is executed directly.

*Work sponsored by SWITCH*

lkiesow committed:
MH-12730, Workflow Index Rebuild Performance

For a workflow-index rebuild Opencast loads and deserializes far more
data that it actually needs. This may even lead to a system needing
hours before it can actually start the index rebuild or for the process
to fail due to the massive amount of required memory.

This patch tries to improve this behavior by requesting the specific
resources directly from the database and deserializing only what is
necessary.

Additionally, it will also avoid sending those messages via message
broker which would only be used for creating log messages. Instead, the
logging is executed directly.

smarquard commented:
Do we actually need the workflow index for anything?

staubesv commented:
@smarquard Which one do you mean? Solr and/or ElasticSearch? My guess would be that the AdminUI uses the data from ElasticSearch and Solr is solely used by the workflow REST endpoint... Lars, what do you think?

smarquard commented:
I mean the solr workflow index. There are endpoints like /workflow/instances.json that require searching, but I don't think those are used by anything as all the searching nowadays is handled by the Admin-UI or External API elasticsearch indexes.

So we could possibly just wire up the REST endpoints for /workflow/ that are still required to the database directly, and drop the unused endpoints, and get rid of the workflow solr index completely.

lkiesow commented on modules/serviceregistry/src/main/java/org/opencastproject/serviceregistry/impl/ServiceRegistryJpaImpl.java:
```diff
@@ -1711,6 +1711,20 @@ public int compare(JpaJob job1, JpaJob job2) {
+  public List<String> getJobPayloads(String operation) throws ServiceRegistryException {
+    EntityManager em = emf.createEntityManager();
+    try {
+      Query query = em.createNativeQuery("SELECT j.payload FROM mh_job j "
```
With Greg's DB renaming patch merged, this is now `oc_job`. I will adjust this pull request…

lkiesow commented:
It's not unlikely that the Solr index could go. I can take a short look at it to evaluate if we can remove it. Will get back to this probably tomorrow.

karendolan commented:
@lkiesow when testing if the workflow index can go, can you test with a very loaded workflow db table (unless db pruning is automatic in 5x) and several running workflows in distributed setup.

lkiesow commented:
I've updated this pull request to fit the new database scheme. It should now be ready to go.

I have also had a short look at the possibility of removing the Solr workflow index and I think it can very likely go. To test this, [I've created a small patch](https://github.com/lkiesow/opencast/commit/9876956506bed2ba530e2cf823249e257e8a339d) to track all index usage and processed a few recordings with that patch included. The result is that the index only gets updated but seems to never be used:

```
2018-03-02 13:57:47,478 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:57:52,409 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:02,498 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:07,440 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:17,592 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:22,495 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:32,597 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:58:37,531 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:23,004 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:25,369 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:30,560 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:35,403 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:50,595 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 13:59:55,453 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 14:00:15,549 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)                                                                                           
2018-03-02 14:00:20,508 | ERROR | (WorkflowServiceImpl:1432) - [>c082ce88] using workflow index: index(workflowInstance)
```

That said, removing this is outside of the scope of this pull request and we probably need to carefully check which REST endpoints would get removed if we rip out the index and which endpoints we want/need to keep and hence re-create.

lkiesow commented:
That said, does anyone mind reviewing this ;-D

gregorydlogan commented on modules/serviceregistry/src/main/java/org/opencastproject/serviceregistry/impl/ServiceRegistryJpaImpl.java:
```diff
@@ -1750,6 +1750,20 @@ public int compare(JpaJob job1, JpaJob job2) {
+  public List<String> getJobPayloads(String operation) throws ServiceRegistryException {
+    EntityManager em = emf.createEntityManager();
+    try {
+      Query query = em.createNativeQuery("SELECT j.payload FROM oc_job j "
```
Is there a reason why this can't be created as a NamedQuery in the JpaJob class?

lkiesow replied:
Good point. There was a reason. But reading this again, it seems like there is none anymore. This should work with a named query as well. I'll try that.

gregorydlogan commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2394,71 +2390,44 @@ public synchronized void cleanupWorkflowInstances(int buffer, WorkflowState stat
-    if (jobs.size() > 0) {
-      logger.info("Populating index '{}' with {} workflows", indexName, jobs.size());
-      final int total = jobs.size();
+    List<String> workflows =  serviceRegistry.getJobPayloads(Operation.START_WORKFLOW.toString());
```
Nitpick: Shouldn't this variable be named payloads?  Alternatively, we should rename the method.

lkiesow replied:
`getJobPayloads(…)` will return the payload of any type of job. It just happens to be that the payload of a `START_WORKFLOW` job is a serialized workflow instance. That's why the name `workflows` for the variable does make sense here while `payloads` makes sense for the method.

gregorydlogan commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2394,71 +2390,44 @@ public synchronized void cleanupWorkflowInstances(int buffer, WorkflowState stat
+
+    final String destinationId = WorkflowItem.WORKFLOW_QUEUE_PREFIX + indexName.substring(0, 1).toUpperCase()
+            + indexName.substring(1);
+    logger.warn("index name: {}", destinationId);
```
Should this be warn?  Or debug?  Caps should be fixed regardless.

lkiesow replied:
I think this is a left-over from testing. This can be removed.

gregorydlogan reviewed (approved):
Approved in general, a few minor questions.

lkiesow committed:
MH-12730, Switch to named query

This patch switches from a native to a named query for database access
to avoid referencing the table directly.

lkiesow commented:
Fixed your comments
